cpp/ql/src/semmle/code/cpp/commons/Buffer.qll

```diff
@@ -71,19 +76,22 @@ int getBufferSize(Expr bufferExpr, Element why) {
         bufferVar.getType().getSize() -
         parentClass.getSize()
     )
-  ) or exists(Expr def |
-    // buffer is assigned with an allocation
-    definitionUsePair(_, def, bufferExpr) and
-    exprDefinition(_, def, why) and
-    isFixedSizeAllocationExpr(why, result)
-  ) or exists(Expr def, Expr e, Element why2 |
-    // buffer is assigned with another buffer
-    definitionUsePair(_, def, bufferExpr) and
-    exprDefinition(_, def, e) and
-    result = getBufferSize(e, why2) and
-    (
+  ) or (
+    // buffer is a fixed size dynamic allocation
+    isFixedSizeAllocationExpr(bufferExpr, result) and
+    why = bufferExpr
+  ) or (
+    // dataflow (all sources must be the same size)
+    forex(Expr def |
+      DataFlow::localFlowStep(DataFlow::exprNode(def), DataFlow::exprNode(bufferExpr)) |
+      result = getBufferSize(def, _)
+    ) and
```

Comment:
This use or `forex` reminds me of the code that was removed in https://git.semmle.com/Semmle/code/commit/90a05cd6fafcd2e179c49f17a5b67bb01de2b57d because it had bad performance on certain pathological snapshots. I'm guessing the same transformation could be needed here.

As I understand it, the performance is quadratic in the number of `def`: for each `def` the implicit `exists` will find one or more `result`, and then each of those `result` will be checked against every other `def` in the implicit `forall`.

Reply:
Nevertheless I've made the change, and it has no noticeable effect on performance (as we'd expect outside of extreme cases).

Reply:
One way to see the the problem is that if you wrote `forall` instead of `forex` here, the compiler would rightfully complain that `result` was not bound: for a `bufferExpr` with no associated `def`, any value of `result` would satisfy the `forall`. This would make `result` take infinitely many values in the sense that the type checker considers `int` to be infinite.

In other words, the `forall` can only _check_ whether a candidate for `result` has the desired properties -- it cannot _generate_ a candidate. When it's joined with an `exists`, the candidates can be generated by `exists`, and then each candidate is checked by the `forall`. If there are _n_ different candidates, and each candidate needs to be compared for equality with the _n_ other candidates, then the run time is _n_^2.

As I'm writing this out, I can see that it would take some quite contrived code to trigger the quadratic behaviour.